server/src/main/java/com/cloud/storage/snapshot/SnapshotManagerImpl.java

```diff
@@ -553,6 +553,25 @@ private void postCreateRecurringSnapshotForPolicy(long userId, long volumeId, lo
         }
     }
 
+    public void markFailedSnapshot(long snapshotId) {
+        Account caller = CallContext.current().getCallingAccount();
+
+        // Verify parameters
+        SnapshotVO snapshotCheck = _snapshotDao.findById(snapshotId);
+
+        if (snapshotCheck == null) {
+            throw new InvalidParameterValueException("unable to find a snapshot with id " + snapshotId);
+        }
+
+        _accountMgr.checkAccess(caller, null, true, snapshotCheck);
+
+        snapshotCheck.setState(Snapshot.State.Error);
+        _snapshotDao.update(snapshotId, snapshotCheck);
+        // Setting removed to prevent record from being deleted by garbage collection.
+        _snapshotDao.remove(snapshotId);
```

Comment:
should this be in a transaction to be sure?

Reply:
ping @Spaceman1984

Reply:
ping @davidjumani @GabrielBrascher @rhtyd the two statements above should be grouped in a transaction. agree?